README.md

```diff
@@ -0,0 +1,84 @@
+# JsonApiDotNetCore MongoDB Repository
+
+[![NuGet version][nuget-image]][nuget-url]
+[![Downloads][downloads-image]][nuget-url]
+[![Build Status](https://travis-ci.com/mrnkr/JsonApiDotNetCore.MongoDb.svg?branch=master)](https://travis-ci.com/mrnkr/JsonApiDotNetCore.MongoDb)
+[![codecov](https://codecov.io/gh/mrnkr/JsonApiDotNetCore.MongoDb/branch/master/graph/badge.svg)](https://codecov.io/gh/mrnkr/JsonApiDotNetCore.MongoDb)
+[![license][license]](https://github.com/mrnkr/JsonApiDotNetCore.MongoDb/blob/master/LICENSE)
+
+[nuget-image]:https://img.shields.io/nuget/v/JsonApiDotNetCore.MongoDb
+[nuget-url]:https://www.nuget.org/packages/JsonApiDotNetCore.MongoDb
+[downloads-image]:https://img.shields.io/nuget/dt/JsonApiDotNetCore.MongoDb
+[license]:https://img.shields.io/github/license/mrnkr/JsonApiDotNetCore.MongoDb
+
+Plug-n-play implementation of `IResourceRepository<TResource, TId>` allowing you to use MongoDb with your `JsonApiDotNetCore` APIs.
+
+## Installation and Usage
+
+### Models
+
+```cs
+public sealed class Book : IIdentifiable<string>
+{
+    [BsonId]
+    [BsonRepresentation(BsonType.ObjectId)]
+    [Attr]
+    public string Id { get; set; }
+
+    [Attr]
+    public string Name { get; set; }
+
+    [BsonIgnore]
+    public string StringId { get => Id; set => Id = value; }
+}
+```
+
+### Controllers
+
+```cs
+public sealed class BooksController : JsonApiController<Book, string>
+{
+    public BooksController(IJsonApiOptions options, ILoggerFactory loggerFactory, IResourceService<Book, string> resourceService)
+        : base(options, loggerFactory, resourceService)
+    {
+    }
+}
+```
+
+### Middleware
+
+```cs
+public class Startup
+{
+    public IServiceProvider ConfigureServices(IServiceCollection services) {
+        services.AddSingleton<IMongoDatabase>(sp =>
+        {
+            var client = new MongoClient(Configuration.GetSection("DatabaseSettings:ConnectionString").Value);
+            return client.GetDatabase(Configuration.GetSection("DatabaseSettings:Database").Value);
+        });
+
+        services.AddScoped<IResourceRepository<Book, string>, MongoEntityRepository<Book, string>>();
+        services.AddJsonApi(options =>
+        {
+            options.Namespace = "api";
+            options.UseRelativeLinks = true;
+            options.IncludeTotalResourceCount = true;
+            options.SerializerSettings.Formatting = Formatting.Indented;
+        }, resources: builder =>
+        {
+            builder.Add<Book, string>();
+        });
+        // ...
+    }
+
+    public void Configure(IApplicationBuilder app)  {
+        app.UseJsonApi();
+        // ...
+    }
+}
+```
+
+## Limitations
+
+- Relations are not supported (yet)
+- Projections are not supported (yet)
```

Comment:
What is the problem when using `fields`?

Reply:
The problem I ran into was that since I was basing myself off of the original `EntityFrameworkCoreRepository` and its dependencies, I found that your `QueryableBuilder` was using the `SelectClauseBuilder` class which depends on some classes exposed by EFCore so I could not bring it into the solution.

I know there is another way I just haven't taken the time to figure it out yet and suggestions are more than welcome.

Reply:
> ...so I could not bring it into the solution.

I don't think that is correct. JsonApiDotNetCore has a dependency on EF Core, and since JsonApiDotNetCore.MongoDb depends on JsonApiDotNetCore, it indirectly also depends on EF Core. NuGet dependencies are transitive, so this should not prevent you from accessing EF Core types. We may split things up in the future, but have no concrete plans for that at the moment.

Easiest solution for now would be to just create a dummy model (you cannot pass `null`), then ensure that `layer.Projection` is always `null` so the dummy model is never used.

```c#
private sealed class DummyModel : IModel
{
    public static IModel Instance { get; } = new DummyModel();

    private DummyModel()
    {
    }

    public IAnnotation FindAnnotation(string name)
    {
        throw new NotImplementedException();
    }

    public IEnumerable<IAnnotation> GetAnnotations()
    {
        throw new NotImplementedException();
    }

    public object this[string name] => throw new NotImplementedException();

    public IEnumerable<IEntityType> GetEntityTypes()
    {
        throw new NotImplementedException();
    }

    public IEntityType FindEntityType(string name)
    {
        throw new NotImplementedException();
    }

    public IEntityType FindEntityType(string name, string definingNavigationName, IEntityType definingEntityType)
    {
        throw new NotImplementedException();
    }
}

// ...

var builder = new QueryableBuilder(source.Expression, source.ElementType, typeof(Queryable), nameFactory, _resourceFactory, _resourceGraph, DummyModel.Instance);
```

This should still give you proper sparse fieldsets in responses, in the sense that only the subset of requested fields is sent back to the client. It just does not optimize the SQL to only fetch the subset of fields. I don't know if that's even possible using MongoDB.

Note that `IncludeBuilder` also depends on an EF Core extension method. If that something MongoDb just ignores or haven't you tried that yet?